File: api/bazaar/report.js
import crypto from "node:crypto";
import { send, readJson, sameOrigin } from "../_lib/http.mjs";
import { readDb, writeDb, currentUser } from "../_lib/bazaar-account.mjs";

const clean = (value, max) => String(value || "").trim().slice(0, max);

export default async function handler(req, res) {
  const user = await currentUser(req);
  if (!user) return send(res, 401, { error: "Entre na sua conta para denunciar." });
  if (req.method !== "POST" || !sameOrigin(req)) return send(res, 405, { error: "Método não permitido." });
  const body = await readJson(req);
  const adId = clean(body.adId, 80);
  const reason = clean(body.reason, 100);
  if (!adId || !reason) return send(res, 400, { error: "Selecione o motivo da denúncia." });
  const db = await readDb();
  db.reports ||= [];
  const duplicate = db.reports.some((report) => report.adId === adId && report.reporterId === user.id && report.status === "aberta");
  if (duplicate) return send(res, 409, { error: "Você já enviou uma denúncia para este anúncio." });
  const report = {
    id: crypto.randomUUID(), adId, title: clean(body.title, 120), seller: clean(body.seller, 40),
    reason, details: clean(body.details, 600), reporterId: user.id, reporter: user.username,
    status: "aberta", createdAt: new Date().toISOString()
  };
  db.reports.push(report);
  await writeDb(db);
  return send(res, 201, { ok: true, id: report.id });
}

File: backend/src/main/java/com/vpertz/reports/Report.java
package com.vpertz.reports;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import java.time.OffsetDateTime;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

/** Denúncia de um anúncio feita por um usuário. */
@Entity
@Table(name = "reports")
@Getter
@Setter
@NoArgsConstructor
public class Report {

    @Id
    @Column(length = 64)
    private String id;

    @Column(name = "ad_id", nullable = false, length = 80)
    private String adId;

    @Column(length = 120)
    private String title;

    @Column(length = 40)
    private String seller;

    @Column(nullable = false, length = 100)
    private String reason;

    @Column(length = 600)
    private String details;

    @Column(name = "reporter_id", length = 64)
    private String reporterId;

    @Column(nullable = false, length = 16)
    private String status = "aberta";

    @Column(name = "created_at", nullable = false)
    private OffsetDateTime createdAt = OffsetDateTime.now();
}
